### ISSUE TITLE
[vue-query]: Different queryCache behavior on SSR for Nuxt3 and Nuxt2

### ISSUE THREAD

**Issue by manuelarling** [TanStack/query#7877]
### Describe the bug

The behavior for `queryClient` using server-side rendering is different and works incorrectly in Nuxt2/Vue2. On Nuxt3/Vue3 the expected behavior occurs.

In Nuxt 3, the `queryCache` is likely reset for each request. This is not the case in Nuxt 2.

**This leads to unexpected behavior in the following example use case.**
When user 1 requests a page, the list of to-dos belonging to user 1 is fetched and displayed. User 2 then accesses the site, but when the page is loaded on the client, the list of to-dos belonging to User 1 is displayed, and the to-dos belonging to User 2 are not displayed until after the refetch on the client page. This creates a privacy issue.

This use case is shown as an example in the two repos. To simulate two different users, the url param is set differently. This is to symbolize the different responses from a server.

### Your minimal, reproducible example
**Nuxt 2:**
https://codesandbox.io/p/devbox/gracious-hermann-lkc3rq?workspaceId=684c7eda-e5ab-41ca-9242-874b5f17c5ae

**Nuxt 3:**
https://codesandbox.io/p/devbox/proud-wood-h3syhd?workspaceId=684c7eda-e5ab-41ca-9242-874b5f17c5ae

### Steps to reproduce

1. Get the nuxt app to running
2. Open the website in one tab with the following url: `/abc`
3. Open the website in a second tab with the following url: `/123`
4. Reload the first tab
5. Reload the second tab

Do this for both repos and you will see that the behavior is different and you will see that the logs in the browser devtools and the terminal are different.

### Expected behavior

The Behaviour on Nuxt2/Vue2 should be the same as on Nuxt3/Vue3.

### How often does this bug happen?

Every time

### **Nuxt 2:**
https://github.com/user-attachments/assets/574214fb-b50e-4af0-84d2-b5a2cfe15588

### **Nuxt 3:**
https://github.com/user-attachments/assets/b5715d5e-4fa0-4023-a646-4331a83cbc8a

### Platform

- OS: [macOS]
- Browser: [Chrome, Firefox]

### Tanstack Query adapter

vue-query

### TanStack Query version

v5.51.21

**Comment by DamianOsipiuk** [comment]
You are right.

This is due to the fact that in Vue2, Vue instance is shared across requests.
Clearing it between requests is not enough, cause there might be couple requests in-flight at the same time.

I believe you might achieve proper per-request separation with slight modification of the plugin and component code.

[Sandbox with the fix](https://codesandbox.io/p/devbox/vue-query-ssr-nuxt-2-forked-8jzcsr?workspaceId=bddce254-c699-46c4-b87a-d54a463560b3&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm02796o00007356eis7uxlwa%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm02796o00002356ehpqadyb3%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm02796o00004356ezhxz2g1e%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm02796o00006356eriypq9ux%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm02796o00002356ehpqadyb3%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm02796nz0001356e0zkskt5z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cm027c6gs000w356d3cnk6tsf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A19%252C%2522startColumn%2522%253A61%252C%2522endLineNumber%2522%253A19%252C%2522endColumn%2522%253A61%257D%255D%252C%2522filepath%2522%253A%2522%252Fplugins%252Fvue-query.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cm02796o00002356ehpqadyb3%2522%252C%2522activeTabId%2522%253A%2522cm027c6gs000w356d3cnk6tsf%2522%257D%252C%2522cm02796o00006356eriypq9ux%2522%253A%257B%2522id%2522%253A%2522cm02796o00006356eriypq9ux%2522%252C%2522activeTabId%2522%253A%2522cm027do17000o356dge0hh4hg%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm02796o00005356ennqd24u8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A2222%252C%2522id%2522%253A%2522cm027do17000o356dge0hh4hg%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm02796o00004356ezhxz2g1e%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm02796o00003356e8blvh07g%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522cm02796o00004356ezhxz2g1e%2522%252C%2522activeTabId%2522%253A%2522cm02796o00003356e8blvh07g%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Plugin code:
```ts
import Vue from "vue";
import { VueQueryPlugin, QueryClient, hydrate } from "@tanstack/vue-query";

export default (context) => {
  // Modify your Vue Query global settings here
  const queryClient = new QueryClient({
    defaultOptions: { queries: { staleTime: 5000 } },
  });

  if (process.server) {
    context.ssrContext.VueQuery = queryClient;
  }

  if (process.client) {
    Vue.use(VueQueryPlugin, { queryClient });

    if (context.nuxtState && context.nuxtState.vueQueryState) {
      console.log(
        "Hydrating state from server",
        context.nuxtState.vueQueryState
      );
      hydrate(queryClient, context.nuxtState.vueQueryState);
    }
  }
};
```

In component, move `queryClient` retrieval to the top and pass `queryClient` to each `useQuery` call as a second parameter:
```ts
const { ssrContext } = useContext();
const queryClient = (ssrContext != null && ssrContext.VueQuery) || useQueryClient()

const { data: data1, suspense } = useQuery({
  queryKey: ["todos"],
  queryFn: fetcher,
}, queryClient);
```

**Comment by manuelarling** [comment]
First of all, thank you for your reply that helped a lot.

I was missing the first information, and then I did some more research. I found out that the problem is also called `cross-request state pollution`. The difference in the two examples in the documentation is that the Nuxt 3 example uses the [`useState`](https://nuxt.com/docs/api/composables/use-state) Composable. This fixes the problem by creating a separate state object for each request.

In any case, I would expect the two examples in the [documentation](https://tanstack.com/query/latest/docs/framework/vue/guides/ssr#using-nuxtjs) to have the same behavior, or at least in the Nuxt 2 example a hint of the potential problem.

Yes, I am looking for a solution that implements a version for creating queryClient per request. Is there perhaps another suggestion, as unfortunately this one requires a lot of refactoring and also attention must be paid in the future.
